examples/tree_3d_dgsem/elixir_mhd_diffusive_convergence.jl

```diff
@@ -0,0 +1,148 @@
+using OrdinaryDiffEq
+using Trixi
+
+###############################################################################
+# semidiscretization of the visco-resistive compressible MHD equations
+
+prandtl_number() = 0.72
+mu_const = 0e-2
+eta_const = 0e-2
+
+equations = IdealGlmMhdEquations3D(5 / 3)
+equations_parabolic = ViscoResistiveMhd3D(equations, mu = mu_const,
+                                          Prandtl = prandtl_number(),
+                                          eta = eta_const,
+                                          gradient_variables = GradientVariablesPrimitive())
+
+volume_flux = (flux_hindenlang_gassner, flux_nonconservative_powell)
+solver = DGSEM(polydeg = 3,
+               surface_flux = (flux_lax_friedrichs, flux_nonconservative_powell),
+               volume_integral = VolumeIntegralFluxDifferencing(volume_flux))
+
+coordinates_min = (-1.0, -1.0, -1.0) # minimum coordinates (min(x), min(y), min(z))
+coordinates_max = (1.0, 1.0, 1.0) # maximum coordinates (max(x), max(y), max(z))
+
+# Create a uniformly refined mesh
+mesh = TreeMesh(coordinates_min, coordinates_max,
+                initial_refinement_level = 2,
+                n_cells_max = 200_000) # set maximum capacity of tree data structure
+
+function initial_condition_constant_alfven(x, t, equations)
+    # Alfvén wave in three space dimensions modified by a periodic density variation.
+    # For the system without the density variations see: Altmann thesis http://dx.doi.org/10.18419/opus-3895.
+    # Domain must be set to [-1, 1]^3, γ = 5/3.
+    omega = 2.0 * pi # may be multiplied by frequency
+    # r = length-variable = length of computational domain
+    r = 2.0
+    # e = epsilon
+    e = 0.02
+    nx = 1 / sqrt(r^2 + 1.0)
+    ny = r / sqrt(r^2 + 1.0)
+    sqr = 1.0
+    Va = omega / (ny * sqr)
+    phi_alv = omega / ny * (nx * (x[1] - 0.5 * r) + ny * (x[2] - 0.5 * r)) - Va * t
+
+    rho = 1.0 + e*cos(phi_alv + 1.0)
+    v1 = -e * ny * cos(phi_alv) / rho
+    v2 = e * nx * cos(phi_alv) / rho
+    v3 = e * sin(phi_alv) / rho
+    p = 1.0
+    B1 = nx - rho * v1 * sqr
+    B2 = ny - rho * v2 * sqr
+    B3 = -rho * v3 * sqr
+    psi = 0.0
+
+#    k = 2*pi
+#    rho = 1.0
+#    v1 = 0
+#    v2 = -e*sin(k*x[1])*sqrt(rho)
+#    v3 = 0
+#    B1 = 1
+#    B2 = e*sin(k*x[1])
+#    B3 = 0
+#    p = 1
+#    psi = 0
+
+    return prim2cons(SVector(rho, v1, v2, v3, p, B1, B2, B3, psi), equations)
+end
+
+
+@inline function source_terms_mhd_convergence_test(u, x, t, equations)
```

Comment:
From the looks of it there source term contribution were determined with computer algebra software (maybe Maple?) and then copied into here. It is hard to parse what exactly is going on. From the initial condition routine it seems that you are using the periodic manufactured solution (MMS) from Altmann's thesis. The energy source term is especially ugly. Would it be possible that you pick a simpler MMS ansatz and then compute the source term again (even by hand if necessary)? This is what I did for the compressible Navier-Stokes MMS test.

Reply:
Yes, that result is from using sympy. I apply the differential operator from the PDE (visco-resistive MHD) on the manufactured solution. There are two test functions. One is a simple Alfven wave in the x-direction and the other is indeed from Altmann's thesis, which is also an Alfven wave. The simpler ansatz is the Alfven wave in x, which fails in the energy term when mu and eta > 0.